xmppserver/src/main/java/org/jivesoftware/openfire/session/LocalSession.java

```diff
@@ -434,44 +451,64 @@ public void deliverRawText(String text) {
 
     @Override
     public void close() {
-        if (conn == null) return;
-        conn.close();
+        lock.lock();
+        try {
+            Optional.ofNullable(conn)
+              .ifPresent(connection ->  connection.close());
+        }finally {
+            lock.unlock();
+        }
     }
 
     @Override
     public boolean validate() {
-        return conn.validate();
+        lock.lock();
+        try {
+            return Optional.ofNullable(conn)
+                    .map(Connection::validate)
+                    .orElse(Boolean.FALSE);
+        }finally {
+            lock.unlock();
+        }
     }
 
     @Override
     public boolean isSecure() {
-        return conn.isSecure();
+        return Optional.ofNullable(conn)
+                .map(Connection::isSecure)
+                .orElse(Boolean.FALSE);
     }
 
     @Override
     public Certificate[] getPeerCertificates() {
-        return conn.getPeerCertificates();
+        return Optional.ofNullable(conn)
+                .map(Connection::getPeerCertificates)
+                .orElse(new Certificate[0]);
     }
 
     @Override
     public boolean isClosed() {
-        return conn.isClosed();
+        return Optional.ofNullable(conn)
+                .map(Connection::isClosed)
+                .orElse(Boolean.TRUE);
     }
 
     @Override
     public String getHostAddress() throws UnknownHostException {
-        if (conn == null) {
+        Connection connection = conn;
```

Comment:
Could be simply

```
return Optional.ofNullable(conn)
    .map(Connection::getHostAddress())
    .orElseThrow(()->new UnknownHostException("Detached session");
```

Reply:
It is not possible, because UnknownHostException is a checked Exception and Connection::getHostAddress() has declerad it. In lambda use it should be like:

```java 
return Optional.ofNullable(conn)
    .map(connection -> {
        try{
            return connection.getHostAddress();
         }catch(UnknownHostException e){
            throw new RuntimeException(e);
        }
     })
    .orElseThrow(()->new UnknownHostException("Detached session");
```

in my opinion this is not better.

Reply:
Good point, well made!